pkg/controller/workspacerouting/workspacerouting_controller.go

```diff
@@ -175,33 +175,109 @@ func (r *ReconcileWorkspaceRouting) Reconcile(request reconcile.Request) (reconc
 		return reconcile.Result{Requeue: true}, err
 	}
 
-	ingressesInSync, err := r.syncIngresses(instance, ingresses)
+	ingressesInSync, clusterIngresses, err := r.syncIngresses(instance, ingresses)
 	if err != nil || !ingressesInSync {
 		reqLogger.Info("Ingresses not in sync")
 		return reconcile.Result{Requeue: true}, err
 	}
 
-	routesInSync, err := r.syncRoutes(instance, routes)
+	routesInSync, clusterRoutes, err := r.syncRoutes(instance, routes)
 	if err != nil || !routesInSync {
 		reqLogger.Info("Routes not in sync")
 		return reconcile.Result{Requeue: true}, err
 	}
 
-	return reconcile.Result{}, r.reconcileStatus(instance, routingObjects)
+	exposedEndpoints, endpointsReady, err := getExposedEndpoints(instance.Spec.Endpoints, clusterIngresses, clusterRoutes)
+	if err != nil {
+		reqLogger.Error(err, "Could not get exposed endpoints for workspace")
+		instance.Status.Phase = workspacev1alpha1.RoutingFailed
+		statusErr := r.client.Status().Update(context.TODO(), instance)
+		return reconcile.Result{}, statusErr
+	}
+
+	return reconcile.Result{}, r.reconcileStatus(instance, routingObjects, exposedEndpoints, endpointsReady)
 }
 
-func (r *ReconcileWorkspaceRouting) reconcileStatus(instance *workspacev1alpha1.WorkspaceRouting, routingObjects solvers.RoutingObjects) error {
+func (r *ReconcileWorkspaceRouting) reconcileStatus(
+	instance *workspacev1alpha1.WorkspaceRouting,
+	routingObjects solvers.RoutingObjects,
+	exposedEndpoints map[string][]workspacev1alpha1.ExposedEndpoint,
+	endpointsReady bool) error {
+
+	if !endpointsReady {
+		instance.Status.Phase = workspacev1alpha1.RoutingPreparing
+		return r.client.Status().Update(context.TODO(), instance)
+	}
 	if instance.Status.Phase == workspacev1alpha1.RoutingReady &&
 		cmp.Equal(instance.Status.PodAdditions, routingObjects.PodAdditions) &&
-		cmp.Equal(instance.Status.ExposedEndpoints, routingObjects.ExposedEndpoints) {
+		cmp.Equal(instance.Status.ExposedEndpoints, exposedEndpoints) {
 		return nil
 	}
 	instance.Status.Phase = workspacev1alpha1.RoutingReady
 	instance.Status.PodAdditions = routingObjects.PodAdditions
-	instance.Status.ExposedEndpoints = routingObjects.ExposedEndpoints
+	instance.Status.ExposedEndpoints = exposedEndpoints
 	return r.client.Status().Update(context.TODO(), instance)
 }
 
+func getExposedEndpoints(
```

Comment:
consider moving endpoints related funcs into a separate file, like to be sync with other helper files -> `resolve_endpoints`. Does it make sense?

Reply:
Moved these functions to `resolve_endpoints.go`